std/os/time.zig

```diff
@@ -0,0 +1,274 @@
+                var freq: i64 = undefined;
+                var err = windows.QueryPerformanceFrequency(&freq);
+                if(err == 0) return error.TimerUnsupported;
+                self.frequency = u64(freq);
+                self.resolution = @divFloor(ns_per_s, self.frequency);
+                var start_time: i64 = undefined;
+                _ = windows.QueryPerformanceCounter(&start_time);
+                self.start_time = u64(start_time);
+            },
+            Os.linux => {
+                var ts: posix.timespec = undefined;
+                var result = posix.clock_getres(monotonic_clock_id, &ts);
+                switch(posix.getErrno(result)) {
+                    0 => {},
+                    posix.EINVAL => return error.TimerUnsupported,
+                    else => unreachable,
+                }
+                self.resolution = u64(ts.tv_sec * ns_per_s + ts.tv_nsec);
+                _ = posix.clock_gettime(monotonic_clock_id, &ts);
+                self.start_time = u64(ts.tv_sec * ns_per_s + ts.tv_nsec);
+            },
+            Os.macosx, Os.ios => {
+                darwin.mach_timebase_info(&self.frequency);
+                self.resolution = @divFloor(self.frequency.numer, self.frequency.denom);
+                self.start_time = darwin.mach_absolute_time();
+            },
+            else => @compileError("Unsupported OS"),
+        }
+        return self;
+    }
+    
+    /// Reads the timer value since start or the last reset in nanoseconds
+    pub fn read(self: &Timer) u64 {
+        var clock = clockNative() - self.start_time;
+        return switch(builtin.os) {
+            Os.windows => @divFloor(clock * ns_per_s, self.frequency),
+            Os.linux => clock,
+            Os.macosx, Os.ios => @divFloor(clock * self.frequency.numer, self.frequency.denom),
+            else => @compileError("Unsupported OS"),
+        };
+    }
+    
+    /// Resets the timer value to 0/now.
+    pub fn reset(self: &Timer) void
+    {
+        self.start_time = clockNative();
+    }
+    
+    /// Returns the current value of the timer in nanoseconds, then resets it
+    pub fn lap(self: &Timer) u64 {
+        var now = clockNative();
+        var lap_time = self.read();
+        self.start_time = now;
+        return lap_time;
+    }
+    
+    
+    const clockNative = switch(builtin.os) {
+        Os.windows => clockWindows,
+        Os.linux => clockLinux,
+        Os.macosx, Os.ios => clockDarwin,
+        else => @compileError("Unsupported OS"),
+    };
+    
+    fn clockWindows() u64 {
+        var result: i64 = undefined;
+        var err = windows.QueryPerformanceCounter(&result);
+        debug.assert(err != 0);
+        return u64(result);
+    }
+    
+    fn clockDarwin() u64 {
+        return darwin.mach_absolute_time();
+    }
+    
+    fn clockLinux() u64 {
+        var ts: posix.timespec = undefined;
+        var result = posix.clock_gettime(monotonic_clock_id, &ts);
```

Comment:
Don't know if it matters but seccomp2 sandboxes can reject `clock_gettime()`, although that's arguably a broken-beyond-reasonableness sandbox.

Reply:
I hadn't considered that a system might allow clock_getres but not clock_gettime. It's possible and therefore should be considered. My first thought is to add another check in start() and only throw the error there, but my understanding is that seccomp could be used to reject syscalls using arbitrary criteria, meaning that just because a syscall succeeded once doesn't mean it will in the future, so the appropriate thing to do would be to make every call that makes a syscall potentially throw an error.

That seems really annoying, but edge cases matter, so I dunno. At the very least I can add a check to start() and make EACESS throw an error instead of unreachable.

Reply:
Actually it turns out seccomp can return an arbitrary errno value. I feel like the best way to handle this is to just use asserts and unreachable because this kind of thing would be roughly on par with using a debugger to insert random values. Open to suggestions though.

Reply:
Sounds reasonable to me.